### ISSUE TITLE
Errors building on Mac Catalina

### ISSUE THREAD

**Issue by nuin** [TileDB-Inc/TileDB-VCF#305]
Hi

I am having a lot of errors when trying to build with `Python` on Catalina. The errors come after this command

```
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include -I/Users/nuin/Projects/VariantsDB/variants_tile/venv/lib/python3.8/site-packages/pyarrow/include -I/Users/nuin/src/TileDB-VCF/dist/include -I/Users/nuin/Projects/VariantsDB/variants_tile/venv/include -I/Users/nuin/.pyenv/versions/3.8.3/include/python3.8 -c src/tiledbvcf/binding/libtiledbvcf.cc -o build/temp.macosx-10.15-x86_64-3.8/src/tiledbvcf/binding/libtiledbvcf.o -std=c++11 -g -O2
```

and end with these errors

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/functional:767:29: note: '::std::greater_equal' declared here
struct _LIBCPP_TEMPLATE_VIS greater_equal : binary_function<_Tp, _Tp, bool>
                            ^
In file included from src/tiledbvcf/binding/libtiledbvcf.cc:1:
In file included from /Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include/pybind11/pybind11.h:45:
In file included from /Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include/pybind11/attr.h:13:
In file included from /Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include/pybind11/cast.h:13:
In file included from /Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include/pybind11/pytypes.h:12:
In file included from /Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.8.egg/pybind11/include/pybind11/detail/common.h:160:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_set:364:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:325:9: error: no member named 'isless' in the global namespace
using ::isless;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:326:9: error: no member named 'islessequal' in the global namespace
using ::islessequal;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:327:9: error: no member named 'islessgreater' in the global namespace
using ::islessgreater;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:328:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:329:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
197 warnings and 13 errors generated.
```

If it helps

```
❯ clang --version
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Any help appreciated.

**Comment by Shelnutt2** [comment]
@nuin We are checking on the error, in the mean time I wanted to point out that we do offer a [prebuilt version with conda](https://docs.tiledb.com/main/solutions/integrations/population-genomics/installation/quick-install)

```
# Install the CLI and shared library
conda install -c conda-forge -c bioconda -c tiledb libtiledbvcf

# Install the Python package
conda install -c conda-forge -c bioconda -c tiledb tiledbvcf-py
```

**Comment by nuin** [comment]
Hi

Thanks @Shelnutt2, I am trying to setup a pure Python, as we might not have Conda available is some machines.

I have the CLI built fine. Is there a way just to install the Python module to interface with the CLI/TileDB file?

**Comment by Shelnutt2** [comment]
@nuin after some research it seems that the problem might be a problem with xcode command line tools on your system. Can you take a look at trying the solution highlights on the swift forms?
https://forums.swift.org/t/is-anyone-else-getting-this-error-when-building-the-compiler-from-master-on-macos/36113/4

**Comment by nuin** [comment]
@Shelnutt2 Deleted the directory as suggested in the thread. I tried with building after a clean and it got stuck on missing 

`ld: library not found for -larrow_python`

By reinstalling command line tools this error passed and I got stuck on the previous one, with the exact same output. I have a conda version installed, I am trying on another machine(s) to check.

Thanks

**Comment by nuin** [comment]
Just for the sake of completeness, different machine, same `clang` but a newer Python than the other one

```
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.9.egg/pybind11/include -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.9.egg/pybind11/include -I/Users/nuin/src/TileDB-VCF/dist/include -I/Users/nuin/Envs/tile/lib/python3.9/site-packages/pyarrow/include -I/Users/nuin/src/TileDB-VCF/dist/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/nuin/Envs/tile/include -I/usr/local/Cellar/python@3.9/3.9.2_2/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c src/tiledbvcf/binding/reader.cc -o build/temp.macosx-11-x86_64-3.9/src/tiledbvcf/binding/reader.o -std=c++11 -g -O2
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.9.egg/pybind11/include -I/Users/nuin/src/TileDB-VCF/apis/python/.eggs/pybind11-2.6.2-py3.9.egg/pybind11/include -I/Users/nuin/src/TileDB-VCF/dist/include -I/Users/nuin/Envs/tile/lib/python3.9/site-packages/pyarrow/include -I/Users/nuin/src/TileDB-VCF/dist/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/nuin/Envs/tile/include -I/usr/local/Cellar/python@3.9/3.9.2_2/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c src/tiledbvcf/binding/writer.cc -o build/temp.macosx-11-x86_64-3.9/src/tiledbvcf/binding/writer.o -std=c++11 -g -O2
clang++ -bundle -undefined dynamic_lookup -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk build/temp.macosx-11-x86_64-3.9/src/tiledbvcf/binding/libtiledbvcf.o build/temp.macosx-11-x86_64-3.9/src/tiledbvcf/binding/reader.o build/temp.macosx-11-x86_64-3.9/src/tiledbvcf/binding/writer.o -L/Users/nuin/src/TileDB-VCF/dist/lib -L/Users/nuin/Envs/tile/lib/python3.9/site-packages/pyarrow -L/Users/nuin/src/TileDB-VCF/dist/lib -L/usr/local/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/sqlite/lib -L/usr/local/opt/tcl-tk/lib -ltiledbvcf -larrow_python -o build/lib.macosx-11-x86_64-3.9/tiledbvcf/libtiledbvcf.cpython-39-darwin.so
ld: library not found for -larrow_python
clang: error: linker command failed with exit code 1 (use -v to see invocation)
error: command '/usr/bin/clang++' failed with exit code 1
```

**Comment by Shelnutt2** [comment]
@nuin We've just put a PR up to fix the python arrow linkage outside of conda. It appears the pypi arrow package lacks the expected symlinks for the arrow libraries from the versioned to unversioned files. That is `libarrow_python.so.300` exists but `libarrow_python.so` does not, which causes the linker to fail. The PR #307 addresses this by always linking against the versioned file. Its not the most ideal solution but it will work and unblock your error for now.

**Comment by nuin** [comment]
Thanks @Shelnutt2

**Comment by nuin** [comment]
Hi

Any news on this? I pulled from master and I am still getting the same error as initially mentioned.

 OTOH, if I remove the Command Line Tools, the build and installation complete fine.

Thanks for the help

**Comment by nuin** [comment]
I don't know if I should open a new ticket, but although build and import work fine now, I have issues running the tests

```
❯ python setup.py pytest
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: invalid command 'pytest'
```

**Comment by aaronwolen** [comment]
Hi, @nuin. I'm glad to hear the building issue has been resolved! 

Regarding the unit tests, assuming pytest is installed and available in your environment, you should be able to run them by running `pytest` within the `apis/python` directory. Let me know if that doesn't work for you.

**Comment by nuin** [comment]
@aaronwolen The pytest issue is that on the documentation is says

<img width="760" alt="image" src="https://user-images.githubusercontent.com/6273/113724024-2b031e80-96af-11eb-91c9-991cdd337403.png">

Running with a system pytest works fine.

Thanks
